ospackage/software_recipes/artifact_repository.go

```diff
@@ -0,0 +1,236 @@
+	// https://commondatastorage.googleapis.com/<bucket>/<object>
+	gsHTTPRegex3 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://(?:commondata)?storage\.googleapis\.com/%s/%s$`, bucketRegex, objectRegex))
+
+	gcsAPIBase = "https://storage.cloud.google.com"
+
+	testStorageClient *storage.Client
+	testHttpClient    *http.Client
+	testFileHandler   FileHandler
+)
+
+func newStorageClient(ctx context.Context) (*storage.Client, error) {
+	if testStorageClient != nil {
+		fmt.Printf("%+#v\n", testStorageClient)
+		return testStorageClient, nil
+	}
+	return storage.NewClient(ctx)
+}
+
+func newHttpClient() *http.Client {
+	if testHttpClient != nil {
+		return testHttpClient
+	}
+	return &http.Client{}
+}
+
+func newFileHandler() FileHandler {
+	if testFileHandler != nil {
+		return testFileHandler
+	}
+	return &OSFileHandler{}
+}
+
+// Takes in a slice of artifacs and dowloads them into the specified directory,
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.name] = path
+	}
+
+	return localNames, nil
+}
+
+func tryGcsRegex(r *regexp.Regexp, url string) (string, string, bool) {
+	matches := gsHTTPRegex1.FindStringSubmatch(url)
+	if len(matches) == 3 {
+		return matches[1], matches[2], true
+	}
+	return "", "", false
+}
+
+func tryTransformGcsUrl(url string) (string, bool) {
+	bucket, object, ok := tryGcsRegex(gsHTTPRegex1, url)
+	if ok {
+		return fmt.Sprintf(gsFormat, bucket, object), true
+	}
+	bucket, object, ok = tryGcsRegex(gsHTTPRegex2, url)
+	if ok {
+		return fmt.Sprintf(gsFormat, bucket, object), true
+	}
+	bucket, object, ok = tryGcsRegex(gsHTTPRegex2, url)
+	if ok {
+		return fmt.Sprintf(gsFormat, bucket, object), true
+	}
+	return "", false
+}
+
+func fetchArtifact(ctx context.Context, a Artifact, directory string) (string, error) {
+	path := path.Join(directory, a.name)
+	u, err := url.Parse(a.url)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", a.url, a.name)
+	}
+	scheme := strings.ToLower(u.Scheme)
+
+	switch scheme {
+	case GCS:
+		err := fetchFromGCS(ctx, a, u, path)
+		if err != nil {
+			return "", err
+		}
+	case Https, Http:
+		gcsLoc, ok := tryTransformGcsUrl(a.url)
+
+		if ok {
+			gcsUrl, err := url.Parse(gcsLoc)
+			if err != nil {
+				return "", err
+			}
+			err = fetchFromGCS(ctx, a, gcsUrl, path)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		err := fetchViaHttp(ctx, a, u, path)
+		if err != nil {
+			return "", err
+		}
+	default:
+		return "", fmt.Errorf("protocol %q found in artifact %q not supported", scheme, a.name)
+	}
+
+	return path, nil
+}
+
+func fetchFromGCS(ctx context.Context, a Artifact, u *url.URL, path string) error {
```

Comment:
np. fetchFrom vs fetchVia inconsistency

Reply:
i don't think this function should take a url.Url but rather explicit bucket, fragment string. parsing the URI string to a url.URL makes sense in calling function, but this function taking url.Url is surprising

Reply:
that makes sense. I changed it to take gcs location info explicitly